
<!DOCTYPE html>
<html class="">
<head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
  <title>STREAM.FeatureSelection GenePattern Module</title>
  <link href="./assets/docstyle.css" media="all" rel="stylesheet">
  <style type="text/css"></style>
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta content="authenticity_token" name="csrf-param">
</head>
<body >

<script>
</script>
	
	<div class="gp-content-header fluid">
		<div class="container-fluid">
			<div class="row">
				<div class="col-md-12">
			        <h4><img src="./assets/gplogo.png"/>Module Documentation</h4>
                                <h1><big>STREAM.VisualizeTrajectories</big></h1>
				</div>
			</div>
		</div>
	</div>
	<div class="container-fluid">
		<div class="row">
			<div class="col-md-12">
				
<br/>
<dl class="dl-horizontal">
	<dt>LSID</dt>
	<dd>urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00401</dd>
	<dt>Author(s)</dt>
	<dd>Huidong Chen, Massachussets General Hospital, wrapped as a module by Ted Liefeld, Mesirov Lab, UCSD School of Medicine.</dd>
	<dt>Contact(s)</dt>
	<dd><p>Algorithm and scientific questions:&nbsp;&lt;Huidong.Chen &nbsp;at mgh dot harvard  dot edu&gt;</p>
<p>Module wrapping issues:&nbsp; Ted Liefeld &nbsp;&lt; jliefeld at cloud dot ucsd dot edu&gt;</p>
</dd>
</dl>


<hr>

<h2>Introduction</h2>

<p>STREAM (Single-cell Trajectories Reconstruction, Exploration And Mapping) is an interactive pipeline capable of disentangling and visualizing complex branching trajectories from both single-cell transcriptomic and epigenomic data.  Within GenePattern STREAM is implemented as a collection of modules that cover the entire STREAM processing pipeline to allow individual steps to be performed interactively for data exploration.</p>

<p>STREAM.VisualizeTrajectories is used to generate the flat tree subway and stream plots that intuitively represent trajectories as linear segments for a computed graph structure. </p>

<h2>Algorithm</h2>
<p>
The tree structure learned in the 3D space (or higher dimensional space), is first approximated by linear segments (each representing a branch) and mapped to a 2D plane based on a modified version of the force-directed layout Fruchterman-Reingold algorithm. We adjust each edge length in order to preserve the lengths of the branches of the original tree. Finally, using both the pseudotime location on the assigned branch and the distance from it in the MLLE space, we map cells to the obtained tree in the 2D plane. Cells are represented as dots and randomly placed to either side of the assigned branches. Each node in the tree indicates one cell state (cell states are sequentially named S0, S1, ... starting from a randomly selected node) and the resulting structure is called **flat tree plot**.
</p><p>
Starting from the flat tree plot and with a designated root or start node, breadth-first search is used to order and arrange nodes and edges horizontally on a 2D plane. Because we preserve the branch lengths of the original tree, the x-axis represents the distance (namely pseudotime) from the start node along the different branches. Cells are then mapped to the obtained structure, and the display of this structure is called a subway map plot. It is created with the same strategy used for the flat tree plot. 
</p><p>
Although these visualizations capture trajectories and branching points, they are not informative on the density and composition of cell types along pseudotime, a common challenge when modeling large datasets. To solve this problem, we develop a trajectory visualization method called the stream plot. This compact representation summarizes cellular developmental trajectories, user-defined annotations, branching points, cell density, and gene expression patterns.
</p><p>
Starting from the subway map plot, for each cell type (if cell labels are provided), using a sliding window approach, we first calculate the number of cells in each window along a developmental branch. To provide smooth transitions around the branching nodes, in those regions the sliding window spans both parent branch and children branches and then proceeds independently on each branch. Then, the numbers of cells in all sliding windows are normalized based on the length of the longest path in the tree. The vertical layout of different branches is optimized by taking into consideration normalized numbers of cells to make sure there will not be overlap between branches. Based on the normalized sliding window values, we first use linear interpolation to construct a set of supporting points. Then the Savitzky-Golay filter (a smoothing filter able to preserve well the signal and avoid oscillations) is applied to create smooth curves based on the set of supporting points. Finally, the obtained curves polygons (one for each cell type) are assembled to form the stream plot. On the stream plot, the length of each branch is the same as in the subway map plot and represents pseudotime, whereas the width is proportional to the number of cells at a given position.

</p>

<h2>References</h2>
<p>H Chen, L Albergante, JY Hsu, CA Lareau, GL Bosco, J Guan, S Zhou, AN Gorban, DE Bauer, MJ Aryee, DM Langenau, A Zinovyev, JD Buenrostro, GC Yuan, L Pinello&nbsp;<a href="https://www.nature.com/articles/s41467-019-09670-4">Single-cell trajectories reconstruction, exploration and mapping of omics data with STREAM. </a> Nature Communications, volume 10, Article number: 1903 (2019)</p>
<p>Nestorowa, S. et al.<a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425"> A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>
<p>Pinello Lab &nbsp; <a href="https://github.com/pinellolab/STREAM">STREAM Github Repository</a></p>

<h2>Input Files</h2>

<ol>
	<li>data file *<br>
 A STREAM pkl file containing an annotated AnnData matrix of gene expression data.</li>
	
</ol>

<h2>Output Files</h2>

<ol>
	<li>&lt;output filename&gt;_stream_result.pkl<br/>
	Output file in STREAM AnnData extended pickle (.pkl) file format suitable for passing to the next step of the STREAM analysis.</li>
	<li>S#&lt;output filename&gt;_stream_plot_&lt;gene_name&gt;.png<bt/>
        Plot of gene expression across the named stream for the named gene. 
        </li>
        <li>S#&lt;output filename&gt;_subwaymap_plot_&lt;gene_name&gt;.png<bt/>
        Subway map plot across the named stream for the named gene.           
        </li>
        <li>&lt;output filename&gt;_flat_tree.png<bt/>
        Tree plot showing all requested gens across the trajectories.           
        </li>



</ol>

<h2>Example Data</h2>

<p>Example data can be downloaded from dropbox: <a href="https://www.dropbox.com/sh/dfqht1ob89ku99d/AACI5ZW3aRuq9MhBfSNS_1O_a?dl=0">Stream Example Data</a><br/>
Ref: Nestorowa, S. et al. <a href="https://www.ncbi.nlm.nih.gov/pubmed/27365425">A single-cell resolution map of mouse hematopoietic stem and progenitor cell differentiation</a>. Blood 128, e20-31 (2016).</p>

<h2>Requirements</h2>

<p>GenePattern 3.9.11 or later (dockerized).</p>


<h2>Parameters</h2>

<table class="table table-striped">
	
	<tbody>
		<tr><td colspan=2 ><h4>General</h4></td></tr>
		<tr>
			<th>Name</th>
			<th>Description</th>
		</tr>
		<tr>
			<td>data file*</td>
			<td>A STREAM pkl file containing an annotated AnnData matrix of gene expression data/td>
		</tr>
		<tr>
			<td>output filename*</td>
			<td>The output filename prefix.</td>
		</tr>
		<tr>
			<td>root   </td>
			<td> The starting node.  </td>
		</tr>
                <tr>
                        <td>percentile dist   </td>
                        <td>Percentile of cells' distances from branches (between 0 and 100) used for calculating the distances between branches of subway map.   </td>
                </tr>
                <tr>
                        <td> preference  </td>
                        <td> The preference of nodes. The branch with speficied nodes are preferred and put on the top part of subway plot. The higher ranks the node have, the closer to the top the branch with that node is. e.g. S3,S4.  </td>
                </tr>
                <tr>
                        <td>color by   </td>
                        <td> 	Specify how to color cells.'label': the cell labels. 'branch': the branch id identifed by STREAM.  </td>
                </tr>
                <tr>
                        <td>subway factor  </td>
                        <td>The factor used to adjust the distances between branches of subway map.   </td>
                </tr>
                <tr>
                        <td>factor num win   </td>
                        <td> The number of sliding windows.  </td>
                </tr>
                <tr>
                        <td> factor min win  </td>
                        <td> The factor used to calculate the sliding window size based on shortest branch.  </td>
                </tr>

                <tr>
                        <td> factor width  </td>
                        <td> The ratio between length and width of stream plot.  </td>
                </tr>



                <tr><td colspan=2>
			<h4>Cell Plots</h4>
                </td></tr>
                <tr>
                        <th>Name</th>
                        <th>Description</th>
                </tr>

                <tr>
                        <td> create cell plots  </td>
                        <td> Create the cell plots, True/False  </td>
                </tr>
                <tr>
                        <td>  log view </td>
                        <td>  If True,the number of cells (the width) is logarithmized when outputing stream plot. </td>
                </tr>
                <tr>
                        <td>factor zoom in   </td>
                        <td>	If "log view" parameter  is True, the factor used to zoom in the thin branches.   </td>
                </tr>
               

 		<tr><td colspan=2>
                        <h4>Gene Plots</h4>
                </td></tr>
                <tr>
                        <th>Name</th>
                        <th>Description</th>
                </tr>
              <tr>
                        <td>   create gene plots</td>
                        <td>  Create the gene plots, True/False. </td>
                </tr>
              <tr>
                        <td>genes   </td>
                        <td> 	Comma delimited list of genes, e.g. Car2,Gata1,Epx.  </td>
                </tr>


                <tr><td colspan=2><h4>Plotting</h4>
                        Parameters controlling the output figures.
                </td></tr>
		<tr>
                        <th>Name</th>
                        <th>Description</th>
                </tr>
                <tr>
                        <td>figure height</td>
                        <td>Figure height as used in matplotlib graphs.  Default=8.</td>
                </tr>
                <tr>
                        <td>figure width</td>
                        <td>Figure width as used in matplotlib plots.  Default=8</td>
                </tr>
  		<tr>
                        <td>figure legend num columns   </td>
                        <td> 	The number of columns that the legend has. </td>
                </tr>

	</tbody>
</table>

<p><span style="color: red;">*</span> - required</p>
			</div>
		</div>
	</div>

	</div>
	<footer class="container fluid gp-footer">
	<div class="fluid text-center">©2006-2019 Regents of the University of California &amp; The Broad Institute of MIT and Harvard</div> 
</footer>



<div id="feedly-mini" title="feedly Mini tookit"></div></body></html>

